### ISSUE TITLE
Comments

### ISSUE THREAD

**Issue by avigad** [avigad/arwm#1]
Remember that the questions are:

- What kinds of automation have you found helpful?
- What kinds of automation do you think would be helpful?

Write as little or as much as you would like. Discussion is welcome.

**Comment by kodyvajjha** [comment]
Scott Morrison's `tidy` tactic never ceases to amaze me when it works. 

I read Scott Morrison say somewhere that `tidy` was inspired by Ganesalingam and Gowers' work, and that there are more ideas in that paper which can lend themselves to more automation.

**Comment by jesse-michael-han** [comment]
Scott Morrison's `library_search` tactic never ceases to amaze me when it works :^)

`finish` still works very well (thanks to Jeremy).

We still lack a tool that can do nested but straightforward induction arguments (possibly involving rewrites with the induction hypotheses). `tidy` can do some, but only if it inducts on the right things and when its call to `simp*` picks up the correct rewrites. That would be useful.

**Comment by jcommelin** [comment]
I think we also need good support (maybe some style conventions suffice) of how to export/cache generated proofs, so that the proof does not have to be rediscovered on every compile of the library. (At the same time, we want to hide the discovered proof from the occasional reader/user of the library.)

**Comment by avigad** [comment]
@jcommelin I agree, when we use expensive search procedures to construct proofs, there are really interesting questions as to what gets stored and what gets checked, and when it gets checked, and how. The SMTCoq has done interesting work on using verified checkers to check certificates efficiently, and Mario Carneiro's current project address this in a foundational way.

The real question I am trying to answer is what it is possible for automation to do and what we want it to do. Once we decide that, it is a secondary question as to how to make it efficient. But you raise a good point that efficiency and convenience are importnat. If the main reason that people don't use automation is that it slows down compilation, that is a good message to convey to developers.

I should also mention, in the notes, that Scott's `tidy` and our `squeeze_simp` begin to address the issue by returning a more efficient tactic script once the solution has been found. I think there is an interesting science to developing compact certificates that can be used to reconstruct a proof efficiently.

**Comment by kim-em** [comment]
> Once we decide that, it is a secondary question as to how to make it efficient.

I think we're already up against a wall here, where the inability to cache generated proofs means that we're not able to use the more powerful tactics, and so maths isn't getting done.

A critical example here is using "heavy" tactics with `auto_param`. Even the solution of having the "heavy" tactic produce a relatively efficient tactic script is not really acceptable, because then the otherwise human-readable source file gets clogged with trivialities.

A couple of times already I've given up on a development because I've realised I'd have to paste in automatically generated proofs, and the ugliness depressed me. :-)

**Comment by kim-em** [comment]
I think there are potentially good solutions. Keeley Hoek made a patch (I'm not sure if it landed in 3.5) that let you do fast serialisation/deserialisation, and hacked the tactic monad in some way that I don't quite understand, so that a `begin ... end` block could effectively cache itself to the file system. We really really need formal language support along these lines.

**Comment by kim-em** [comment]
Addressing your original question, I would like to see:
* `list_applies`, that suggests possible `apply ...` statements for the current goal
* `list_rewrites`, similarly for `rw ...`
* `hint`, that runs through a bunch of tactics (`refl`, `simp`, ...) and/or hammers (`finish`, `linarith`, `tidy`, `omega`, ...) and prints a trace message about what it found
  * the ability to just have `hint` running continuously in the background!
* a GUI term inspector, that lets you expand selected implicit/universe arguments, and that lets you run tactics on subexpressions (i.e. producing `conv` blocks for you)

**Comment by kbuzzard** [comment]
`ring` and `norm_num` have been essential for me; indeed I think both were made into Lean tactics by Mario as a result of my constantly moaning on the Lean chat that I couldn't do things which a high school kid would find trivial.

**Comment by avigad** [comment]
Thanks, everyone, for the input. It is really helpful.

A few days ago, I was CC'ed on an e-mail exchange discussing a paper that had recently appeared on arXiv (https://arxiv.org/pdf/1907.07794.pdf). Frank Pfenning made the following comment:

> ... sometimes I am worried that too much automation is a bad thing, because it can mask the flaws in what one is doing. There are many examples in the study of logics, but let me use programming instead.  Start with your favorite terrible programming language (Python happens to be a convenient target, or C++).  Now, sure, you can reduce the number of bugs with automatic program analysis, the smarter the better.  But you would nevertheless still be better off in a statically typed functional language with a decent module system.  You prop up something that would be better left to wither and die.

I like this quote (which I am sharing with his permission) because it is a nice expression of a view held by some in the programming language community, namely, that automation is the wrong response to a problem that is better solved by using / developing better languages.

To be clear: he was talking about verifying software. Whether or not he is right, there is the additional question as to whether how well the claim transfers to verifying mathematics. But I think the quote helps explain some of the divergent attitudes we see.

**Comment by avigad** [comment]
In case anyone is still following along, I just added detailed notes on an experiment with Isabelle:

https://github.com/avigad/arwm/tree/master/isabelle_experiments

**Comment by immler** [comment]
In my opinion, the kind of automation that Isabelle's auto provides is extremely useful, in particular the fact that auto (as opposed to sledgehammer) makes partial progress. 

But I can also relate to Frank's quote:
Often times when I use auto (especially with fact collections like {continuous,tendsto,derivative}_intros) it is only to discharge side conditions that would perhaps better be left to a (more expressive) type system.

That is probably why (as you described in your very nice "preliminary notes") Johannes uses much less automation in Lean.
But perhaps Lean's elaborator/type checker can be seen as automation, too?

**Comment by fpvandoorn** [comment]
I like to use automation that works on specialized problems, and will be predictable in these scenarios. Things like `norm_num`, `library_search`, `omega`, `ring`, `norm_cast`, ...

I also use `simp` a lot. It's less specialized and predictable, but is useful in a very wide number of cases.

I'm less in favor of using powerful tactics, where I cannot judge in advance whether they will be able to solve the goal. Tactics like `tidy`, `finish` and `eblast`. They tend to mostly succeed when I already have a relatively good idea of how to prove the goal myself, and then writing a more explicit proof is nicer, in my opinion.

**Comment by jcommelin** [comment]
> I'm less in favor of using powerful tactics, where I cannot judge in advance whether they will be able to solve the goal. Tactics like `tidy`, `finish` and `eblast`. They tend to mostly succeed when I already have a relatively good idea of how to prove the goal myself, and then writing a more explicit proof is nicer, in my opinion.

I kind of agree, unless these powerful hammers solve 99% of their goals using some `auto_param`. Like functoriality proofs in category_theory. When it works (and is fast) the result is very slick and readable.

**Comment by fpvandoorn** [comment]
Yes, I agree with that. They are very useful if they are applied on problems where they almost always succeed.

**Comment by angeliki-kou** [comment]
Regarding Isabelle/HOL's automation, see Section 5  here https://www.researchgate.net/publication/334549483_FORMALISING_MATHEMATICS_-IN_PRAXIS_A_MATHEMATICIAN%27S_VERY_FIRST_EXPERIENCES_WITH_ISABELLEHOL

**Comment by avigad** [comment]
Let me add that Angeliki tells me that she uses Sledgehammer frequently and finds it very helpful. Here are some examples of proofs of hers where you can see calls to `metis` that were generated by Sledgehammer:

https://www.isa-afp.org/browser_info/current/AFP/Transcendence_Series_Hancl_Rucki/Transcendence_Series.html
https://www.isa-afp.org/browser_info/current/AFP/Octonions/Octonions.html

And I just met with Yong Kiam Tan, a student of André Platzer at CMU, who learned to use Isabelle this summer by formalizing the Lasalle invariant principle and theorems about Lyaponov stability. He thinks Sledgehammer is great. He basically wrote structured proofs and called on Sledgehammer to fill in the details, adding intermediate statements as necessary. He says that it worked very well.
